# Mini Projeto Avaliativo - Análises de  (AED)
### Base Varejo
#### Aluno: Anildo Santos Turma: Análises de dados com Python T3

In [187]:
# importando Biblioteca
import pandas as pd

In [188]:
# Criando DataFrame
# Ao abrir Os arquivos vir que são separados por ponto e vírgula e não somente por vírgula
# utilizei o sep=";" e se tiver acentos o encording="utf-8" vai garantir que sejam lidos corretamente.

df = pd.read_csv("Base Varejo.csv",sep=";",encoding="utf-8")

df

,DATA,CO_ID,CL_ID,CL_GENERO,CL_EC,CL_FHL,CL_SEG,PR_ID,PR_CAT,PR_NOME,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13
0,01/02/2019,1000,534,M,4,1,C,67,BEBIDAS,REFRIGERANTE GUARANA,NaN,NaN,NaN,NaN
1,01/02/2019,1000,534,M,4,1,C,70,BEBIDAS,REFRIGERANTE OUTROS,NaN,NaN,NaN,NaN
2,01/02/2019,1000,534,M,4,1,C,178,HIGIENE,LENCO UMEDECIDO,NaN,NaN,NaN,NaN
3,01/02/2019,1000,534,M,4,1,C,4,ALIMENTOS,ABACAXI,NaN,NaN,NaN,NaN
4,01/02/2019,1000,534,M,4,1,C,175,LIMPEZA,LIMPADOR MULTIUSO,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
829995,19/08/2022,919822,155,F,2,0,B,183,ALIMENTOS,KETCHUP,NaN,NaN,NaN,NaN
829996,19/08/2022,919822,155,F,2,0,B,56,ALIMENTOS,QUEIJO MUSSARELA,NaN,NaN,NaN,NaN
829997,19/08/2022,919822,155,F,2,0,B,227,ALIMENTOS,ARROZ,NaN,NaN,NaN,NaN
829998,19/08/2022,919822,155,F,2,0,B,214,ALIMENTOS,CEBOLA,NaN,NaN,NaN,NaN


In [189]:
# Verificando como cada Dado analisado se encontra para as devidas correções.
print("Numeros de Registros", df.shape[0])
print("Numeros de Colunas", df.shape[1])
df.dtypes

Numeros de Registros 830000
Numeros de Colunas 14


DATA               str
CO_ID            int64
CL_ID            int64
CL_GENERO          str
CL_EC            int64
CL_FHL           int64
CL_SEG             str
PR_ID            int64
PR_CAT             str
PR_NOME            str
Unnamed: 10    float64
Unnamed: 11    float64
Unnamed: 12    float64
Unnamed: 13    float64
dtype: object

# Analisando a Tabela

In [190]:
# Pecorre toda minha tabela e verifica onde há valores nulos e vazios , e me mostra a quantidade os somando

df.isnull().sum()

DATA                0
CO_ID               0
CL_ID               0
CL_GENERO           0
CL_EC               0
CL_FHL              0
CL_SEG              0
PR_ID               0
PR_CAT              0
PR_NOME             0
Unnamed: 10    830000
Unnamed: 11    830000
Unnamed: 12    830000
Unnamed: 13    830000
dtype: int64

In [191]:
# Verificando quantas linhas duplicadas há no arquivo e me mostra a quantidade

print("Linhas_Duplicadas" , df.duplicated().sum())

Linhas_Duplicadas 96553


In [192]:
# Verificando se há produtos com categoria vazia, onde ela tava usando o #N/D como um nulo desfarçado

print("Produtos_vazios (#N/D):", (df["PR_CAT"]=="#N/D").sum())

Produtos_vazios (#N/D): 3650


In [193]:
# Verificando se as datas estão corretas utilizando o datetime que vai converter texto em data  caso esteja
# E utilizando o isna() para contar as falhas,  e o errors="coerce" para transformar um vaalor invalido em vazio

df_data_correta = pd.to_datetime(df["DATA"], format= "%d/%m/%Y",errors="coerce")

df_data_correta.isna().sum()

np.int64(0)

# Limpeza de Dados

In [194]:
# Criando uma cópia dos Dados para uma futura comparação mantendo o Dado original intacto, e fazer a limpeza toda nessa cópia
# Uma vez que já se sabe o que vai ser Limpo.

df_Limpeza_Dados = df.copy()

In [195]:
# Tratando Colunas Vazias eliminando elas

df_Limpeza_Dados = df.drop(columns=["Unnamed: 10", "Unnamed: 11", "Unnamed: 12", "Unnamed: 13"])

df_Limpeza_Dados

,DATA,CO_ID,CL_ID,CL_GENERO,CL_EC,CL_FHL,CL_SEG,PR_ID,PR_CAT,PR_NOME
0,01/02/2019,1000,534,M,4,1,C,67,BEBIDAS,REFRIGERANTE GUARANA
1,01/02/2019,1000,534,M,4,1,C,70,BEBIDAS,REFRIGERANTE OUTROS
2,01/02/2019,1000,534,M,4,1,C,178,HIGIENE,LENCO UMEDECIDO
3,01/02/2019,1000,534,M,4,1,C,4,ALIMENTOS,ABACAXI
4,01/02/2019,1000,534,M,4,1,C,175,LIMPEZA,LIMPADOR MULTIUSO
...,...,...,...,...,...,...,...,...,...,...
829995,19/08/2022,919822,155,F,2,0,B,183,ALIMENTOS,KETCHUP
829996,19/08/2022,919822,155,F,2,0,B,56,ALIMENTOS,QUEIJO MUSSARELA
829997,19/08/2022,919822,155,F,2,0,B,227,ALIMENTOS,ARROZ
829998,19/08/2022,919822,155,F,2,0,B,214,ALIMENTOS,CEBOLA


In [196]:
# Tratando Produtos Nulos trabalhando encima da cópia utilizando o replace para trocar o texto #N/D por Produto Vazio,
#  e o isnull pra confirmar se tudo ta 100% sem nulos e conta-los com sum()

df_Limpeza_Dados["PR_CAT"] = df_Limpeza_Dados["PR_CAT"].replace("#N/D", "Sem Categoria")

df_Limpeza_Dados.isnull().sum()


DATA         0
CO_ID        0
CL_ID        0
CL_GENERO    0
CL_EC        0
CL_FHL       0
CL_SEG       0
PR_ID        0
PR_CAT       0
PR_NOME      0
dtype: int64

In [197]:
# Eliminando as Duplicadas utilizando o len que vai me contar quantas linhas tem duplicadas antes de ser removida guardando pra ser comparada depois 
# Enquanto isso o drop_duplicate vai limpando as linhas repetidas mantendo a ocorrência inicial ou seja a primeira linha de cada uma 
# E foi feita uma subtração de quantas linhas sairam em ( qtd_linhas_antes-len(df_limpeza_Dados) )

qtd_linhas_antes  = len(df_Limpeza_Dados)
print("Linhas_Iniciais",len(df_Limpeza_Dados))

df_Limpeza_Dados = df_Limpeza_Dados.drop_duplicates()
print("Linhas_Duplicadas: ",qtd_linhas_antes-len(df_Limpeza_Dados))

Linhas_Iniciais 830000
Linhas_Duplicadas:  96553


In [198]:
# Tratando datas com formato em texto utilizando o to.datetime para formato de data correta 

df_Limpeza_Dados["DATA"] = pd.to_datetime(df_Limpeza_Dados["DATA"], format="%d/%m/%Y")
df_Limpeza_Dados.dtypes

DATA         datetime64[us]
CO_ID                 int64
CL_ID                 int64
CL_GENERO               str
CL_EC                 int64
CL_FHL                int64
CL_SEG                  str
PR_ID                 int64
PR_CAT                  str
PR_NOME                 str
dtype: object

In [199]:
print(df_Limpeza_Dados["DATA"])

0        2019-02-01
1        2019-02-01
2        2019-02-01
3        2019-02-01
4        2019-02-01
            ...    
829991   2022-08-19
829993   2022-08-19
829994   2022-08-19
829998   2022-08-19
829999   2022-08-19
Name: DATA, Length: 733447, dtype: datetime64[us]


# Estatísticas dos Filhos dos clientes

In [200]:
# Criaçãao de DataFrame

df_filhos_clientes = df_Limpeza_Dados["CL_FHL"]
df_filhos_clientes


0         1
1         1
2         1
3         1
4         1
         ..
829991    0
829993    0
829994    0
829998    0
829999    0
Name: CL_FHL, Length: 733447, dtype: int64

In [201]:
# Analizando a coluna de filhos e aplicando as estatísticas nos dados 
# -count()
# -mean()
# -median()
# -round()
# -mode()
# -min()
# -max() 

df_filhos_clientes = df_Limpeza_Dados["CL_FHL"]

print("Contagem:", df_filhos_clientes.count())
print("Média:", round(df_filhos_clientes.mean(), 2))
print("Mediana:", df_filhos_clientes.median())
print("Desvio padrão:", round(df_filhos_clientes.std(), 2))
print("Moda:", df_filhos_clientes.mode()[0])
print("Mínimo:", df_filhos_clientes.min())
print("Máximo:", df_filhos_clientes.max())

Contagem: 733447
Média: 1.15
Mediana: 0.0
Desvio padrão: 1.42
Moda: 0
Mínimo: 0
Máximo: 4


In [202]:
# Cauculando o quantis dos filhos dos clientes mostrando quantos filhos correspondente a 25% , 50% , 75%

porcentagem = df_filhos_clientes.quantile([0.25,0.5,0.75])

porcentagem

0.25    0.0
0.50    0.0
0.75    2.0
Name: CL_FHL, dtype: float64

# Agrupando Dados

In [203]:
# Agrupando Clientes por gênero e contando  quantos clientes existem por categoria de Produto e quanttas linhass cada grupo tem com o size()

df_Limpeza_Dados.groupby("CL_GENERO").size()

CL_GENERO
F    382427
M    351020
dtype: int64

In [204]:
# Agrupamento por categoria de produto e vendo qual categoria mais vende ordenando do maior para o menor com o sort_values

df_Limpeza_Dados.groupby("PR_CAT").size().sort_values(ascending=False)


PR_CAT
ALIMENTOS        384197
HIGIENE          137702
LIMPEZA          128632
BEBIDAS           38264
PET               28553
ACESSORIOS        12871
Sem Categoria      3228
dtype: int64

# Salvando novo Arquivo CSV

In [205]:
# Salvando o DataFrame em um novo arquivo CSV
# sep=() - > mantém os separadores originais do arquivo (Base Varejo.csv)
# index() - > não salvará linhas automáticas gerando coluna extra
# encodding="utf-8" -> Garante o salvamento dos acentos de modo correto

df_Limpeza_Dados.to_csv("df_Limpeza_Dados.csv",sep=";", index=False, encoding="utf-8")

print("Arquivo salvo com Sucesso: df_Limpeza_Dados.csv")

Arquivo salvo com Sucesso: df_Limpeza_Dados.csv


## Conclusões

1- A base de dados tinha 4 colunas totalmente  vazias, havia sobra de espaço ( ";;;") no arquivo original e 96,553 duplicadas que foram removidas por não trazer nenhuma informmação útil ou representando a mesma venda lançada mais de uma vez.

2- A coluna categoria do produto havia registros com "#N/D" que estavam como texto desfarçados, que foram retirados e tratados como "Sem Categoria" em vez de apaga-los por completo, presevando-o como dado válido como venda.

3- A coluna data precisou ser convertida de texto pra `datetime` para análises por período.

4- foi verificado que a coluna de filhos (`CL_FHL`) não possui valores nulos, e todos os 733.447 registros possuem informaçôes preenchidas.

5- A categoria que maais veennddeu foi `ALIMENTOS`, E O Gênero `F` foi o que mais comprou na Base Varejo.

6- Todo Cliente tem em média `1,15filhos` com concentração na maioria em 0 e 2 filhos.